### PR TITLE
(#8526): Upgrade leveldown from 5.6.0 to 6.1.1

### PR DESCRIPTION
**Pull request by an9949an**
Leveldown needs to be upgraded to support Apple M1. https://github.com/pouchdb/pouchdb/issues/8526

### CODE REVIEW

**Comment by alxndrsn** [comment]
This PR is currently identical to:

* https://github.com/pouchdb/pouchdb/pull/8727/files
* https://github.com/pouchdb/pouchdb/pull/8666/files
* https://github.com/pouchdb/pouchdb/pull/8529/files

It looks like the only CI failures are when testing with webpack.

**Comment by an9949an** [comment]
@alxndrsn yes, I saw. But all this PRs outdated and looks abandoned with all CI logs cleared. I just saw that something fails and this is it.

> It looks like the only CI failures are when testing with webpack.

Yes, but very strange. After a brief investigation I see that it fails at `acorn` parse of `node_modules/abstract-level/abstract-chained-batch.js` on the spread operator. I have no idea what's going on for now. Probably some dependencies incompatibility.

I will try to take a look a bit later. Or maybe somebody has ideas why it happens...

**Comment by alxndrsn** [comment]
My guess is it's related to the ecmascript version that webpack is expecting.

**Comment by an9949an** [comment]
@alxndrsn It fails because of `bundle-collapser` which used at `bin/build-utils.js:53`:
```js 
bundler = browserify(addPath(pkgName, filepath), opts)
  .transform('es3ify')
  .plugin('bundle-collapser/plugin');
```
`bundle-collapser` depends on `falafel 2.2.5` which depends on `acorn ^7.1.1` (resolved as 7.4.1).
Both `bundle-collapser`(last commit 4 years ago) and `falafel` (last commit 3 years ago) look abandoned.

Removing of `bundle-collapser` fixing it at least locally. But I don't sure what to do with this, probably somebody with project knowledge needed here. 

Also for some strange reason it doesn't fail during the normal build (exact same line executed during the `npm run build` many times, even with exact same file) and fails when `CLIENT=firefox npm run test-webpack`.

**Comment by garethbowen** [comment]
@an9949an It looks like all bundle-collapser does is help minify the output a little. Could you run with and without and see what the difference in bundle size is?

**Comment by an9949an** [comment]
@garethbowen Here is the differences

<details>
  <summary>With collapser</summary>
  
</details>

<details>
  <summary>Without collapser</summary>
  
</details>

Sometimes it's near nothing (~1kb difference), sometimes not. The worst case is something like `pouchdb.localstorage.min.js`: 288kb against 297kb (~3% increase)

**Comment by garethbowen** [comment]
That seems acceptable to me. Would you mind updating this PR to remove those libs so we can ensure the tests pass?

**Comment by an9949an** [comment]
@garethbowen It's already removed in this PR.

Also one more topic to think about: the `level` package. Right now we have `level:6.0.1` which depends on `leveldown:^5.4.0`. The latest version is `level:8.0.0` which depends on `leveldown:6.1.1` (I updated `leveldown` to `6.1.1` in direct dependencies). So, should we update the level package as well maybe?

**Comment by garethbowen** [comment]
@an9949an Let's run this as it is and then raise a new PR for updating `level` otherwise we risk making everything too complicated again...

File: bin/build-utils.js [hunk 1]
```diff
@@ -50,7 +50,6 @@ function doBrowserify(pkgName, filepath, opts, exclude) {
     } else {
       bundler = browserify(addPath(pkgName, filepath), opts)
         .transform('es3ify')
```
Comment (garethbowen) [inline bin/build-utils.js:52]: This needs a semicolon to pass linting...

Comment (an9949an) [reply bin/build-utils.js]: Yes, sorry, fixed.

**Comment by garethbowen** [comment]
@an9949an Unfortunately there are quite a few test failures. Can you run the tests locally and see if you can reproduce the issue?

**Comment by an9949an** [comment]
@garethbowen I can't reproduce locally.

As I can see from the job logs it runs fine on `node:16` (the same locally) and fails on `node:14`. I tried to run on the exact same node version and locally no any errors. Also `firefox` version the same. The only difference is distro. Job use `ubuntu:22.04.3`, I use `fedora:38`. The same test which fails by `20s` timeout complete locally during `2s` with no errors.

Another interesting aspect: as I can see in logs, the job fails on the different tests on the first and second try but both by 20s timeout error. Also the same test (_suite2 test.replication_events.js-http-http -> #4627 Test no duplicate changes in live replication_) which failed by timeout during the second retry was complete in 2108ms during the first one.

So, it looks like something random and maybe env related. Did you see something like that before?

**Comment by garethbowen** [comment]
The tests are sometimes unreliable... I'm re-running the failed builds to check.

**Comment by an9949an** [comment]
@garethbowen one fail less this time :) I tried to run this test locally and it's also fine. But I don't completely sure what is the right way to run the same `indexeddb` test. I use this command:  `CLIENT=firefox ADAPTERS=indexeddb npm test`. Is it the right one?

**Comment by garethbowen** [comment]
@an9949an I re-ran again and it passed.

@alxndrsn I think we're ok to merge this now. What do you think? Are these failures within the bounds of normal?

**Comment by alxndrsn** [comment]
I suspect the correct approach is to remove `es3ify` as well - both of them seem old & unmaintained, and neither of them seem to cope with spread operators:

```
> require('es3ify').transform('(...args) => console.log(...args)')
'(...args) => console.log(...args)'
```

and https://github.com/browserify/bundle-collapser/issues/25

I'd vote for removing them in a separate PR.

**Comment by garethbowen** [comment]
@alxndrsn So that's a yes for merging this as is?

**Comment by alxndrsn** [comment]
> @alxndrsn So that's a yes for merging this as is?

No, I'd rather see bundler-collapser removed first, in a separate PR - to me it's removal seems incidental to upgrading leveldown.

**Comment by alxndrsn** [comment]
The spread operator is used elsewhere in the pouchdb source (and likely in other dependencies), so I don't understand _why_ removing `bundle-collapser` fixes things.  Ideally this would be understood before changing the build process.

**Comment by an9949an** [comment]
> The spread operator is used elsewhere in the pouchdb source (and likely in other dependencies), so I don't understand _why_ removing `bundle-collapser` fixes things. Ideally this would be understood before changing the build process.

@alxndrsn very good question. Unfortunately it fails somewhere inside the build libs without any meaningful output. It was even complicated to figure out where exactly it failed. I did it by adding logs to the `acorn` package source code. But why only this place and why firefox test only - I have no idea honestly.

**Comment by an9949an** [comment]
@alxndrsn for now I prepared PR with removing of `es3ify` and `bundle-collapser`.

> Ideally this would be understood before changing the build process

Yes, but as I can see from your comment at least `es3ify` do nothing meaningful. Spread operator introduced in ECMAScript 6. If `es3ify` don't remove spread it mean that result is still incompatible with ES3. Am I wrong?

**Comment by alxndrsn** [comment]
> Yes, but as I can see from your comment at least `es3ify` do nothing meaningful. Spread operator introduced in ECMAScript 6. If `es3ify` don't remove spread it mean that result is still incompatible with ES3. Am I wrong?

I think you're correct, and I'm 100% in favour of removing `es3ify` immediately.

**Comment by alxndrsn** [comment]
@an9949an does this PR pass with `bundle-collapser`, but with `es3ify` removed?

**Comment by an9949an** [comment]
> @an9949an does this PR pass with `bundle-collapser`, but with `es3ify` removed?

@alxndrsn yes, it fails with the same `SyntaxError: Unexpected token`.

As I can see `bundle-collapser` is also not something very important. It just replaces `require('path/to/something')` with `require(1)` to save a bit of bytes.

**Comment by garethbowen** [comment]
@an9949an I've merged the other PR which removes the outdated libs: https://github.com/pouchdb/pouchdb/pull/8847

Can you rebase this one so it's just bumping leveldown?

**Comment by an9949an** [comment]
@garethbowen done

**Comment by garethbowen** [comment]
@an9949an Looks good to me. Waiting for the build to pass.

I haven't tested this as I don't have an M1 - have you tested this build to ensure it actually fixes the issue?

**Comment by an9949an** [comment]
@garethbowen  I don't have it unfortunately. But we tested it on our project by replacing leveldown binaries to ARM one from leveldown 6.1 and everything worked fine on apple M1.

**Review by garethbowen — APPROVED** [review]
Let's do it.

**Pull request merged**